### PR TITLE
Rewarded Video

### PR DESCRIPTION
**Pull request by gracebarsny**
**General**
- Added video rewards: 100 coins when ad button pressed on main menu, dog return from scouting when the whistle button is pressed on its profile
- Please verify that the whistle SFX plays after the ad for dog return is completed!
 
**Prefabs**
- AdoptedDogProfile: renamed a few sub-gameobjects, fixed a spelling error, added a ScoutingButton component to toggle between collar and whistle icons

**Resources**
- AudioList.json: Added whistle SFX

**Scripts**
- AdController: added methods to offer a reward after an ad is played
- AdReward, CoinsAdReward, DogReturnAdReward: new scripts to control ad rewards
- PPTimer: removed unused library, fixed spelling error
- AdoptedDogProfile: added ref to scouting button
- ScoutingButton: new script to handle scouting button on adopted dog profile
- DogCollarSlot: fixed styles (comments and spacing)
- DogHomeSlot: overrode accessor in parent DogSlot class -- home slots no longer reassign a dog's occupied slot (this was causing bugs clearing the collar slot after a dog was brought back by whistle)
- DogSlot: marked dog accessor as virtual, made _dog variable protected
- MainMenu: made ad button reward 100 coins upon video completion
- Dog: moved MonoBehaviourExtended Overrides towards the top, added ReturnFromScout method

### CODE REVIEW

**Comment by imann24** [comment]
Getting a null ref when I try to select a dog from the DogBrowser:

```
NullReferenceException: Object reference not set to an instance of an object
PPUIController.showDogProfile (.Dog dog) (at Assets/Scripts/UI/PickupPup/Core/PPUIController.cs:112)
PPUIController.handleDogSlotClicked (.Dog dog) (at Assets/Scripts/UI/PickupPup/Core/PPUIController.cs:104)
PPUIController.handlePPDogEvent (PPEvent gameEvent, .Dog dog) (at Assets/Scripts/UI/PickupPup/Core/PPUIController.cs:96)
EventController.Event (PPEvent gameEvent, .Dog dog, Boolean isCallback) (at Assets/Scripts/Event/EventController.cs:238)
DogSlot.ExecuteClick () (at Assets/Scripts/UI/PickupPup/DogSlot/DogSlot.cs:170)
UIButton.executeClick () (at Assets/Scripts/UI/Core/Button/UIButton.cs:127)
UnityEngine.Events.InvokableCall.Invoke (System.Object[] args) (at /Users/builduser/buildslave/unity/build/Runtime/Export/UnityEvent.cs:153)
UnityEngine.Events.InvokableCallList.Invoke (System.Object[] parameters) (at /Users/builduser/buildslave/unity/build/Runtime/Export/UnityEvent.cs:634)
UnityEngine.Events.UnityEventBase.Invoke (System.Object[] parameters) (at /Users/builduser/buildslave/unity/build/Runtime/Export/UnityEvent.cs:769)
UnityEngine.Events.UnityEvent.Invoke () (at /Users/builduser/buildslave/unity/build/Runtime/Export/UnityEvent_0.cs:53)
UnityEngine.UI.Button.Press () (at /Users/builduser/buildslave/unity/build/Extensions/guisystem/UnityEngine.UI/UI/Core/Button.cs:35)
UnityEngine.UI.Button.OnPointerClick (UnityEngine.EventSystems.PointerEventData eventData) (at /Users/builduser/buildslave/unity/build/Extensions/guisystem/UnityEngine.UI/UI/Core/Button.cs:44)
UnityEngine.EventSystems.ExecuteEvents.Execute (IPointerClickHandler handler, UnityEngine.EventSystems.BaseEventData eventData) (at /Users/builduser/buildslave/unity/build/Extensions/guisystem/UnityEngine.UI/EventSystem/ExecuteEvents.cs:52)
UnityEngine.EventSystems.ExecuteEvents.Execute[IPointerClickHandler] (UnityEngine.GameObject target, UnityEngine.EventSystems.BaseEventData eventData, UnityEngine.EventSystems.EventFunction`1 functor) (at /Users/builduser/buildslave/unity/build/Extensions/guisystem/UnityEngine.UI/EventSystem/ExecuteEvents.cs:269)
UnityEngine.EventSystems.EventSystem:Update()
```

File: Assets/Scripts/AdsAndIAP/AdReward.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿/*
+ * Author(s): Grace Barrett-Snyder
+ * Description: An in-game reward for watching ads. 
+ */
+
+using k = PPGlobal;
+
+public abstract class AdReward {
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/AdReward.cs]: Newline for bracket

File: Assets/Scripts/AdsAndIAP/CoinsAdReward.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+using k = PPGlobal;
+
+public class CoinsAdReward : AdReward
+{
+    PPGameController gameController;
+    int amount;
+
+    public CoinsAdReward(int amount = 100)
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/CoinsAdReward.cs]: This default amount should be captured w/in a constant in `PPGlobal`

Comment (imann24) [inline Assets/Scripts/AdsAndIAP/CoinsAdReward.cs]: Also confused as to why this constructor takes an `amount` argument if the `amount` class value is set based on a tuning value

File: Assets/Scripts/AdsAndIAP/DogReturnAdReward.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿/*
+ * Author(s): Grace Barrett-Snyder
+ * Description: An in-game reward for watching an ad that immediately returns a dog from scouting. 
+ */
+
+using k = PPGlobal;
+
+public class DogReturnAdReward : AdReward {
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/DogReturnAdReward.cs]: Bracket

File: Assets/Scripts/AdsAndIAP/DogReturnAdReward.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/DogReturnAdReward.cs]: Extra newline

File: Assets/Scripts/UI/PickupPup/DogProfile/ScoutingButton.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
+﻿/*
+ * Author(s): Grace Barrett-Snyder
+ * Description: An in-game reward for watching an ad that gives the player coins.
```
Comment (imann24) [inline Assets/Scripts/UI/PickupPup/DogProfile/ScoutingButton.cs]: Fix description

File: Assets/Scripts/UI/PickupPup/DogProfile/ScoutingButton.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+    public void OnPress()
+    {
+        if(dog.IsScouting)
+        {
+            AdController.GetInstance.WatchAd(new DogReturnAdReward(dog));
+        }
+        else
```
Comment (imann24) [inline Assets/Scripts/UI/PickupPup/DogProfile/ScoutingButton.cs]: I think it's worth excluded the else and just describing in the comment the work still to be done

File: Assets/Scripts/World/PickupPup/Core/Dog.cs [hunk 7]
```diff
@@ -218,6 +218,32 @@ CurrencyData redeemableGift
...
+        {
+            scoutingTimer.Init();
+            setupTimer(scoutingTimer);
+        }
+    }
+
+    protected override void fetchReferences()
+    {
+        base.fetchReferences();
+        gameController = PPGameController.GetInstance;
```
Comment (imann24) [inline Assets/Scripts/World/PickupPup/Core/Dog.cs]: These calls already occur in `MonoBehaviourExtended`

File: Assets/Scripts/AdsAndIAP/AdController.cs [hunk 8]
```diff
@@ -56,13 +55,26 @@ IEnumerator ShowAdWhenReady()
             {
                 switch (result)
                 {
+                    case ShowResult.Finished:
+                        Debug.Log("Advertisement Finished");
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/AdController.cs]: Remove debug statement

File: Assets/Scripts/AdsAndIAP/AdController.cs [hunk 9]
```diff
@@ -56,13 +55,26 @@ IEnumerator ShowAdWhenReady()
...
             }
-
         });
     }
 
+    void offerReward(AdReward reward)
+    {
+        if(reward != null)
+        {
+            Debug.Log("Offer a reward to the player");
```
Comment (imann24) [inline Assets/Scripts/AdsAndIAP/AdController.cs]: Remove debug statement

**Comment by gracebarsny** [comment]
@imann24 Ready for another review! I really appreciated the detail on the last one.

**Comment by imann24** [comment]
Sounds good. I'm heading to bed for the night, but I'll review tomorrow morning

File: Assets/Scripts/UI/PickupPup/DogSlot/DogHomeSlot.cs [hunk 10]
```diff
@@ -5,5 +5,21 @@
...
+        {
+            // A chance to handle any cleanup
+            if (_dog != null)
+            {
+                handleChangeDog(_dog);
+            }
+            // Do not reassign slot
+            _dog = value;
+        }
+    }
```
Comment (imann24) [inline Assets/Scripts/UI/PickupPup/DogSlot/DogHomeSlot.cs]: Add newline

**Comment by imann24** [comment]
@gracebarsny Feel free to merge yourself whenever you fix that one missing bracket

**Pull request merged**